### PR TITLE
Develop wake steering lookup table toolset

### PR DESCRIPTION
**Pull request by misi9170**
This pull request adds tools to WHOC for the design of lookup table (LUT)-based wake steering controllers for wind farms. Visualization tools are also added, and the lookup-based_wake_steering example is updated to demonstrate design capabilities. In particular, the following tools have been added, most of which require an instantiated [`FlorisModel`](https://github.com/NREL/floris/blob/main/floris/floris_model.py)of the wind farm as an input to design procedures:
- [x] Basic wake steering lookup table (using the [`YawOptimizationSR` method](https://github.com/NREL/floris/blob/main/floris/optimization/yaw_optimization/yaw_optimizer_sr.py))
- [x] Wake steering under wind direction uncertainty (uses `YawOptimizationSR` with the [`UncertainFlorisModel`](https://github.com/NREL/floris/blob/main/floris/uncertain_floris_model.py)
- [x] Wake steering with static rate limits (based on code from a stale repository)
- [x] Computing hysteresis regions for wake steering (based on code from a stale repository)
- ~Spline approximations of the optimal wake steering yaw angles (based on code from a stale repository)~
- [x] Conversion of the LUT into a numpy interpolator (bringing over from [FLASC](https://github.com/NREL/flasc/blob/main/flasc/utilities/lookup_table_tools.py), whose version will likely be deprecated)

Will also need to build tests for each design tool and update examples and documentation.
- [x] Tests
- [x] Docstrings
- [x] User documentation
- [x] Examples

### Acknowldegements
Many of the tools incorporated in this PR are not "new" and are being ported over from other repositories in a consolidation effort. Contributors to those feeder repositories include (but there are likely others that I'm missing):
- @Bartdoekemeijer
- @paulf81 
- @ejsimley

### CODE REVIEW

**Comment by misi9170** [comment]
I've now extended the lookup-based_wake_steering_florisstandin example to include a hysteresis option, and updated the documentation accordingly.

**Comment by misi9170** [comment]
After discussions with @paulf81 and @ejsimley , I have decided to leave out the spline approximator. This tool would have allowed users to specify offset lookup tables with a reduced number of datapoints, which could be helpful in some circumstances, but is likely to be more nuanced than a general tool would afford. However, I'm making a note of it here in case there is future interest in such a tool (which we can then port over from another older repository).

**Comment by misi9170** [comment]
I've now added documentation, and this PR is ready for review. The documentation may finally be reorganized some (and API documentation added), but I decided to leave that for a separate pull request.

**Review by ejsimley — CHANGES_REQUESTED** [review]
Thanks for consolidating these lookup table functions here! This looks good overall. I left several comments where I thought we could make some functions more general and improve the interpolation. We can discuss those ideas more if you'd like before making any changes.

File: whoc/design_tools/wake_steering_design.py [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+        # Copy highest wind speed / TI solutions to 999.0
+        ids_to_copy_ub = points[:, col] == np.max(points[:, col])
+        points_copied = np.array(points[ids_to_copy_ub, :], copy=True)
+        values_copied = np.array(values[ids_to_copy_ub, :], copy=True)
+        points_copied[:, col] = 999.0  # Upper bound
+        points = np.vstack([points, points_copied])
+        values = np.vstack([values, values_copied])
+
+    # Now create a linear interpolant for the yaw angles
+    interpolant = LinearNDInterpolator(points=points, values=values, fill_value=np.nan)
```
[the remark this answers is not in the document]
Comment (misi9170) [reply whoc/design_tools/wake_steering_design.py]: Thanks for the suggestion @ejsimley---I've now made this change and added some extra tests for the interpolator.

File: whoc/design_tools/wake_steering_design.py [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+        ws_min=ws_min,
+        ws_max=ws_max,
+        ti=ti,
+    )
+
+    fmodel.set(wind_data=wind_rose)
+
+    ufmodel = UncertainFlorisModel(
+        configuration=fmodel.core.as_dict(),
+        wd_std=wd_std,
```
[the remark this answers is not in the document]
Comment (misi9170) [reply whoc/design_tools/wake_steering_design.py:118]: Good suggestion---I've now done this. I considered allowing the user to pass an instantiated UncertainFlorisModel but decided against it, figuring that if users want full control over the design process, they should use the `YawOptimizerSerialRefine` directly and avoid these `build_xx_wake_steering_lookup_table()` functions. However, I need to document that better (that these are a starting point, but users who want more control should use FLORIS' methods directly)

File: whoc/design_tools/wake_steering_design.py [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+        offsets_limited_lr[:, i, :, :] = offsets_limited_lr[:, i-1, :, :] + delta_yaw
+    offsets_limited_rl = offsets_array.copy()
+    for i in range(len(wd_array)-2, -1, -1):
+        delta_yaw = offsets_limited_rl[:, i, :, :] - offsets_limited_rl[:, i+1, :, :]
+        delta_yaw = np.clip(delta_yaw, -wd_rate_limit*wd_step, wd_rate_limit*wd_step)
+        offsets_limited_rl[:, i, :, :] = offsets_limited_rl[:, i+1, :, :] + delta_yaw
+    offsets_array = (offsets_limited_lr + offsets_limited_rl) / 2
+
+    # Apply ws rate limits (increasing ws)
+    for j in range(1, len(ws_array)):
```
[the remark this answers is not in the document]
Comment (misi9170) [reply whoc/design_tools/wake_steering_design.py:190]: No good reason! I think I was feeling that wind speed and TI were more "linear", and wind direction more "circular", but that is not actually a good reason to only apply the rate limit one way---I've now changed that!

File: examples/lookup-based_wake_steering_florisstandin/compare_yaw_offset_designs.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+        ws_max=ws_max,
+        ti_resolution=ti_resolution,
+        ti_min=ti_min,
+        ti_max=ti_max,
+        minimum_yaw_angle=minimum_yaw_angle,
+        maximum_yaw_angle=maximum_yaw_angle,
+    )
+
+    print("\nApplying rate limits to simple lookup table (3 deg/deg).")
+    df_opt_rate_limited = wsd.apply_static_rate_limits(
```
Comment (paulf81) [inline examples/lookup-based_wake_steering_florisstandin/compare_yaw_offset_designs.py:123]: I'm ok leaving this in, but I think in general we've found this generally not too preferable approach versus rate limiting/filtering in the time domain.  Maybe we just add some comments to the effect that this was an idea but no longer preferred

**Review by paulf81 — COMMENTED** [review]
I think we should add a README.md file to examples/lookup-based_wake_steering_florisstandin

File: whoc/design_tools/wake_steering_design.py [hunk 5]
```diff
@@ -0,0 +1,630 @@
...
+from floris import FlorisModel, UncertainFlorisModel, WindRose, WindTIRose
+from floris.optimization.yaw_optimization.yaw_optimizer_sr import YawOptimizationSR
+from scipy.interpolate import interp1d, RegularGridInterpolator
+
+
+def build_simple_wake_steering_lookup_table(
+    fmodel: FlorisModel,
+    wd_resolution: float = 5,
+    wd_min: float = 0,
+    wd_max: float = 360,
```
Comment (paulf81) [inline whoc/design_tools/wake_steering_design.py:12]: Totally up to you, but I would do 360.0 to insist on it being a float.  But looking around I think either is fine

File: whoc/design_tools/wake_steering_design.py [hunk 6]
```diff
@@ -0,0 +1,630 @@
...
+    Args:
+        fmodel (FlorisModel): An instantiated FlorisModel object.
+        wd_resolution (float, optional): The resolution of the wind direction in degrees.
+            Defaults to 5.
+        wd_min (float, optional): The minimum wind direction in degrees. Defaults to 0.
+        wd_max (float, optional): The maximum wind direction in degrees. Defaults to 360.
+        ws_resolution (float, optional): The resolution of the wind speed in m/s.
+            Defaults to 1.
+        ws_min (float, optional): The minimum wind speed in m/s. Defaults to 8.
+        ws_max (float, optional): The maximum wind speed in m/s. Defaults to 8.
```
Comment (paulf81) [inline whoc/design_tools/wake_steering_design.py:39]: I was looking into the code and note that use add .001 to ws_max, so it could be worth noting that the arrays will be defined with ws_max inclusive, just to make the behavior clear.  So in the default 
```
>>> np.arange(8,8.001,1)
array([8.])
```

File: whoc/design_tools/wake_steering_design.py [hunk 7]
```diff
@@ -0,0 +1,630 @@
...
+    yaw_opt = YawOptimizationSR(
+        fmodel=ufmodel,
+        minimum_yaw_angle=minimum_yaw_angle,
+        maximum_yaw_angle=maximum_yaw_angle,
+    )
+
+    return yaw_opt.optimize()
+
+
+def apply_static_rate_limits(
```
Comment (paulf81) [inline whoc/design_tools/wake_steering_design.py:153]: Made the comment elsewhere but wanted to find a way this method is probably not preferred to putting limits in the time domain.  This is forcing the controller to sub-optimal angles even if the wind direction changes slowly (freely admit this was I think my idea originally...)

File: whoc/design_tools/wake_steering_design.py [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+    wd_step = (wind_directions[1]-wind_directions[0])
+    if (wind_directions[0] - wd_step < 0) & (wind_directions[-1] + wd_step >= 360):
+        offsets = np.concatenate((offsets, offsets[0:1, :, :, :]), axis=0)
+        wd_centers = wind_directions + 0.5 * wd_step
+        wind_directions = np.concatenate((wind_directions, [wind_directions[-1] + wd_step]))
+    else:
+        wd_centers = wind_directions[:-1] + 0.5 * wd_step
+    
+    # Define function that identifies hysteresis zones
+    switching_idx = np.argwhere(np.diff(offsets, axis=0) >= yaw_rate_threshold*wd_step)
```
[the remark this answers is not in the document]
Comment (paulf81) [reply whoc/design_tools/wake_steering_design.py:294]: Makes sense to me

File: whoc/design_tools/wake_steering_visualization.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+        ti_plot = np.unique(df_opt.turbulence_intensity)
+        if ti_plot.size > 1:
+            raise ValueError(
+                "Multiple turbulence intensities present in df_opt. Must specify ti_plot."
+            )
+
+    ws_array = np.unique(df_opt.wind_speed)
+    wd_array = np.unique(df_opt.wind_direction)
+
+    # Construct array of offets
```
Comment (paulf81) [inline whoc/design_tools/wake_steering_visualization.py:59]: typo

**Review by paulf81 — APPROVED** [review]
Left only minor optional comments, looks really great, thanks for the thorough job, this is very useful to have all consolidated and cleaned up!!

File: whoc/design_tools/wake_steering_design.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (misi9170) [reply whoc/design_tools/wake_steering_design.py]: I've now update the methodology a bit for hysteresis, which addresses this comment. I'll write a more detailed explanation of the change as a stand-alone comment soon.

**Comment by misi9170** [comment]
> I think we should add a README.md file to examples/lookup-based_wake_steering_florisstandin

@paulf81 Thanks for the suggestion here. In WHOC, I've so far avoided putting readme's in each of the example folder and instead provided information about each example in docs/examples.md. Is that suitable?

**Comment by misi9170** [comment]
Hi @paulf81, @ejsimley 

It took me a long time to come back to addressing your suggestions; sorry about that. If I could bother you for a rereview, I'd appreciate that. It's not particularly urgent. See the comments above for specifics about the changes I made to address your larger comments.

File: whoc/design_tools/wake_steering_design.py [hunk 10]
```diff
@@ -0,0 +1,668 @@
...
+                        max(hysteresis_wds[i_h][1], hysteresis_wds[i_h+1][1])
+                    )
+                    # Remove next region
+                    hysteresis_wds.pop(i_h+1)
+                else:
+                    i_h += 1
+            # Check the final region, which could be reversed
+            if ((hysteresis_wds[-1][0] > hysteresis_wds[-1][1])
+                or (hysteresis_wds[0][0] > hysteresis_wds[0][1])):
+                if hysteresis_wds[-1][1] > hysteresis_wds[0][0]:
```
Comment (ejsimley) [inline whoc/design_tools/wake_steering_design.py:348]: Is the objective here to consolidate the first and last regions if they overlap around 0 degrees? Trying some scenarios, here are some examples where the regions wouldn't get combined:
`hysteresis_wds = [(3, 7), (359, 3)]`
`hysteresis_wds = [(359, 1), (357, 359)]`

I think this could be changed to `hysteresis_wds[-1][1] >= hysteresis_wds[0][0]`

File: whoc/controllers/lookup_based_wake_steering_controller.py [hunk 11]
```diff
@@ -64,6 +103,16 @@ def wake_steering_angles(self):
             yaw_offsets = np.diag(interpolated_angles)
             yaw_setpoint = (np.array(wind_directions) - yaw_offsets).tolist()
 
+        # Apply hysteresis
+        if self.hysteresis_dict is not None:
+            for t in range(self.n_turbines):
+                for zone in self.hysteresis_dict["T{:03d}".format(t)]:
+                    if zone[0] < wind_directions[t] < zone[1]:
```
Comment (ejsimley) [inline whoc/controllers/lookup_based_wake_steering_controller.py:110]: Does this need to be changed to account for zones that include 0 degrees (e.g., zone[0] == 355, zone[1] == 3)

**Review by ejsimley — CHANGES_REQUESTED** [review]
@misi9170, thanks for making these changes to the wake steering tools! It will be helpful to be able to include TI in the optimizations now. I also appreciate being able to select a particular TI when plotting the yaw offsets.

I like the changes to the hysteresis zone definition, defining a minimum zone width and using the yaw rate threshold to identify hysteresis directions. I just had some comments about how overlapping hysteresis regions get combined. I think some edge cases where regions wrap around 0 degrees aren't getting handled correctly. I included a couple ideas, but there are still some cases that wouldn't get combined correctly. If you agree that the code should be changed, we can talk about these ideas more. 

Otherwise, I just had a couple optional comments in other places.

File: whoc/design_tools/wake_steering_visualization.py [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+                if ti >= ti_array[0] and ti <= ti_array[-1]:
+                    offsets_list.append(
+                        offsets_all[(df_opt.wind_speed == ws) & (df_opt.turbulence_intensity == ti)]
+                    )
+
+    if ax is None:
+        _, ax = plt.subplots(1, 1)
+
+    for offsets in offsets_list:
+        ax.plot(wd_array, offsets, color=color, linestyle=linestyle, alpha=alpha, label=label)
```
[the remark this answers is not in the document]
Comment (misi9170) [reply whoc/design_tools/wake_steering_visualization.py:179]: My preference on this is to let the user call `ax.legend()` on the returned output `ax`. That allows the user to more easily specify other arguments to `legend()` so they control how it looks

File: whoc/design_tools/wake_steering_design.py [hunk 13]
```diff
@@ -0,0 +1,668 @@
...
+            Defaults to 0.06.
+        minimum_yaw_angle (float, optional): The minimum (inclusive) allowable misalignment in
+            degrees. Defaults to 0.0.
+        maximum_yaw_angle (float, optional): The maximum (inclusive) allowable misalignment in
+            degrees. Defaults to 25.0.
+
+    Returns:
+        pd.DataFrame: A yaw offset lookup table.
+    """
+    wind_rose = create_uniform_wind_rose(
```
[the remark this answers is not in the document]
Comment (misi9170) [reply whoc/design_tools/wake_steering_design.py:56]: Yeah, I'm not sure what the protocol is on this. I tend to define the "main" functionality near the top of the file, and helpers later on, but you're right that it means that the function appears "before" it is defined.

**Comment by misi9170** [comment]
Thanks @ejsimley ! I had thought I'd figured out combining the hysteresis zones around the 0/360 divide, but apparently not---I'll look into that and let you know when I think I've fixed it

File: whoc/design_tools/wake_steering_design.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (ejsimley) [reply whoc/design_tools/wake_steering_design.py:56]: That works too. No need to change anything.

File: whoc/design_tools/wake_steering_design.py [hunk 14]
```diff
@@ -0,0 +1,668 @@
...
+            lb = wrap_360(wd_switch_point - min_zone_width/2)
+            ub = wrap_360(wd_switch_point + min_zone_width/2)
+            hysteresis_wds.append((lb, ub))
+
+        # Consolidate regions
+        for _ in range(len(hysteresis_wds)): # Outer loop to handle multiple possible wrap merges
+            i_h = 0
+            while i_h < len(hysteresis_wds)-1:
+                if (hysteresis_wds[i_h][1] >= hysteresis_wds[i_h+1][0]
+                    and hysteresis_wds[i_h][0] <= hysteresis_wds[i_h+1][1]):
```
[the remark this answers is not in the document]
Comment (misi9170) [reply whoc/design_tools/wake_steering_design.py:335]: This last edge case you suggested is a tricky one! I've broken out the consolidation of the hysteresis zones into it's own function, partly just so that I can test it better. It's passing the "easier" tests, but it does indeed fail on this last one (two zones, both crossing the 360/0 divide). I'll see if I can figure that case out.

[the remark this answers is not in the document]
Comment (misi9170) [reply whoc/design_tools/wake_steering_design.py]: Ok, finally I got this one! Had to reorganize the zone combining code some.

**Comment by misi9170** [comment]
Hi @ejsimley ,

After some swearing and soul-searching, I revamped the zone combination algorithm and I think it's now working as intended. I guess I can't be sure that there isn't some edge case that could slip by, I think this is working pretty well now. I've added various possible test scenarios to `test_consolidate_hysteresis_zones()` in tests/wake_steering_design_test.py.

If that all looks good to you, I think I can proceed to merging this in.
